src/platform/intel/cavs/lib/clk.c

```diff
@@ -145,24 +129,67 @@ static inline void set_cpu_current_freq_idx(int freq_idx, bool release_unused)
 {
 	select_cpu_clock(freq_idx, true);
 }
+#endif /* !(CONFIG_CAVS_LPRO_ONLY) */
 
 static void platform_clock_low_power_mode(int clock, bool enable)
 {
+#if CONFIG_CAVS_USE_LPRO_IN_WAITI
+	int current_freq_idx = get_current_freq_idx(clock);
+	int freq_idx = *cache_to_uncache(&active_freq_idx);
+
+	if (enable && current_freq_idx != CPU_LPRO_FREQ_IDX)
+		/* LPRO requests are fast, but requests for other ROs
+		 * can take a lot of time. That's why it's better to
+		 * not release active clock just for waiti,
+		 * so they can be switched without delay on wake up.
+		 */
+		select_cpu_clock(CPU_LPRO_FREQ_IDX, false);
+	else if (!enable && current_freq_idx != freq_idx)
+		select_cpu_clock(freq_idx, false);
+#endif /* CONFIG_CAVS_USE_LPRO_IN_WAITI */
 }
-#endif
 
-void platform_clock_on_waiti(void)
+static int clock_platform_set_cpu_freq(int clock, int freq_idx)
 {
+	set_cpu_current_freq_idx(freq_idx, true);
+	return 0;
 }
 
-void platform_clock_on_wakeup(void)
+void platform_clock_on_waiti(void)
 {
+/* do nothing for LPRO_ONLY mode */
+#if !(CONFIG_CAVS_LPRO_ONLY)
+	/* perform the possible dynamic clock switching */
+	if (*cache_to_uncache(&active_freq_idx) != CPU_LPRO_FREQ_IDX &&
+	    !pm_runtime_is_active(PM_RUNTIME_DSP, PLATFORM_MASTER_CORE_ID))
```

Comment:
Clock switching may base on pm_runtime settings in short term. However, the pm_runtime calls should be prepared for unbalanced pm_gate ipc requests, by replacing ref-counter with a bitmap (one bit per client, could be reduced to a single bool/bit for host->pm_gate use only in short term).

Reply:
> Clock switching may base on pm_runtime settings in short term. However, the pm_runtime calls should be prepared for unbalanced pm_gate ipc requests, by replacing ref-counter with a bitmap (one bit per client, could be reduced to a single bool/bit for host->pm_gate use only in short term).

Thanks for comments, @mmaka1 . I just created a new PR to handle that issue, please help to review:
https://github.com/thesofproject/sof/pull/3252

Reply:
This looks over complex. Is there any reason why we dont store the clock status in uncache memory by default ? You are also checking PM and cached states twice which is not atomic.

Reply:
@keyonjie I'm still seeing the "complex" version here ? Have you pushed the new version ?

Reply:
also far easier to flow if you do
```c
freq_idx = uncache(&active_freq_idx);
pm_is_active = pm_runtime_is_active(PM_blah, MASTER_CORE);

if (pm_is_active) {
    if (freq_idx != LPRO)
        do_this;
} else {
    if (freq_idx != HPRO)
        do_that;
}
```

Reply:
> also far easier to flow if you do
> 
> ```c
> freq_idx = uncache(&active_freq_idx);
> pm_is_active = pm_runtime_is_active(PM_blah, MASTER_CORE);
> 
> if (pm_is_active) {
>     if (freq_idx != LPRO)
>         do_this;
> } else {
>     if (freq_idx != HPRO)
>         do_that;
> }
> ```

sure, let me add these variables for easier understanding.

BTW, have you got my changes that adding the hold of lock for the atomic issue you were worrying about, is that OK to you?

Reply:
yes, GH all working now.

Reply:
@lgirdwood BTW, I will prefer this as it is simple enough for understanding after 2 new variables introduced:
```
	/* perform clock switching if the PM active status not aligned with the clock frequency */
	if (!pm_is_active && freq_idx != CPU_LPRO_FREQ_IDX)
		set_cpu_current_freq_idx(CPU_LPRO_FREQ_IDX, true);
	else if (pm_is_active && freq_idx != CPU_HPRO_FREQ_IDX)
		set_cpu_current_freq_idx(CPU_HPRO_FREQ_IDX, true);
```

Reply:
But you are checking the same logic twice, someone using a code scanning tool will just come along an fix it.